plugins/carlabase/carla.cpp

```diff
@@ -299,14 +306,69 @@ void CarlaInstrument::saveSettings(QDomDocument& doc, QDomElement& parent)
         return;
 
     QDomDocument carlaDoc("carla");
-
     if (carlaDoc.setContent(QString(state)))
     {
         QDomNode n = doc.importNode(carlaDoc.documentElement(), true);
         parent.appendChild(n);
     }
-
     std::free(state);
+
+	for (uint32_t i=0; i < floatModels.count(); ++i)
+	{
+		QString idStr = CARLA_SETTING_PREFIX + QString::number(i);
+		floatModels[i]->saveSettings( doc, parent, idStr );
+	}
+	
+}
+
+void CarlaInstrument::refreshParams()
+{
+	if (fDescriptor->get_parameter_count != nullptr &&
+		fDescriptor->get_parameter_info  != nullptr &&
+		fDescriptor->get_parameter_value != nullptr &&
+		fDescriptor->set_parameter_value != nullptr)
+	{
+		uint32_t param_count = fDescriptor->get_parameter_count(fHandle);
+		if (param_count > 0)
+        {
+			clearKnobModels();
+			floatModels.reserve(param_count);
+            for (uint32_t i=0; i < param_count; ++i)
+            {
+				// https://github.com/falkTX/Carla/tree/master/source/native-plugins source/native-plugins/resources/carla-plugin
+                const NativeParameter* paramInfo(fDescriptor->get_parameter_info(fHandle, i));
+                float param_value = fDescriptor->get_parameter_value(fHandle, i);
+
+				// Get parameter name
+                QString name = "_NO_NAME_";
+				if (paramInfo->name != nullptr){
+					name = paramInfo->name;
+				}
+
+				// Create new model for the knob.
+				floatModels.push_back(new FloatModel(param_value,0.0f,1.0f,0.001f,this,name));
+
+				// Load settings into model.
+				QString idStr = CARLA_SETTING_PREFIX + QString::number(i);
+				floatModels[i]->loadSettings( settingsElem, idStr );
+
+				// Connect to signal dataChanged to knobChanged function.
+				connect( floatModels[i], &FloatModel::dataChanged, [=]() { knobModelChanged(i); });
```

Comment:
Connecting slots to functors is not supported in Qt4. You have two choices:
- Use normal `SIGNAL()`/`SLOT()` and use [`sender()`](http://doc.qt.io/archives/qt-4.8/qobject.html#sender) to detect the sender(what we already do for VSTs)
- Keep this syntax and switch to `master` branch which has dropped the support for Qt4

Reply:
I also think you should use `Qt::DirectConnection` because automated changes will be handled in the UI thread if you use the normal connection type. It may prevent automation working correctly when exporting.